### PR TITLE
[#30946] YSQL: Support traceparent comment at the query start and end

### PR DESCRIPTION
**Pull request by egladysh**
Extends OTel traceparent extraction to accept the comment at the end of a query (SELECT 1 /*traceparent='...'*/) in addition to the existing leading position. Only leading and trailing block comments are checked, avoiding any ambiguity with /* patterns inside SQL string literals. The leading comment takes precedence when both are present.

---

Phorge: [D52273](https://phorge.dev.yugabyte.com/D52273)

### CODE REVIEW

File: src/postgres/src/backend/tcop/postgres.c [hunk 1]
```diff
@@ -7859,39 +7859,73 @@ disable_statement_timeout(void)
...
+		content = pos + YB_TRACEPARENT_COMMENT_DELIMITERS_LEN;
+		comment_end = strstr(content, "*/");
+		if (comment_end)
+		{
+			result = YbGetTraceparentFromTraceContext(content, comment_end - content,
+													  traceparent_out);
+			if (result != YB_TRACEPARENT_NO_FIELD)
+				return result;
+		}
+	}
```
[the remark this answers is not in the document]
Comment (egladysh) [reply src/postgres/src/backend/tcop/postgres.c:7891]: accepted limitation.

File: src/postgres/src/backend/tcop/postgres.c [hunk 2]
```diff
@@ -7859,39 +7859,73 @@ disable_statement_timeout(void)
...
-	if (pos[0] != '/' || pos[1] != '*')
-		return YB_TRACEPARENT_NO_COMMENT;
-
-	/* Skip the comment start delimiters "/ *". */
-	pos += YB_TRACEPARENT_COMMENT_DELIMITERS_LEN;
-
+	/* Check for a trailing block comment. */
+	end = query + strlen(query);
+	while (end > query && isspace((unsigned char) end[-1]))
+		end--;
```
[the remark this answers is not in the document]
Comment (egladysh) [reply src/postgres/src/backend/tcop/postgres.c]: Fixed

File: src/postgres/src/backend/tcop/postgres.c [hunk 3]
```diff
@@ -7867,39 +7859,74 @@ disable_statement_timeout(void)
...
+
  */
 static YbTraceparentResult
 YbExtractTraceParentFromComment(const char *query, char *traceparent_out)
 {
-	const char *pos = query;
-	const char *comment_end;
-
-	/* Skip leading whitespace. */
+	const char		   *pos;
```
Comment (IshanChhangani) [inline src/postgres/src/backend/tcop/postgres.c]:
```suggestion
	const char *pos;
```
Can we follow the postgres way of spacing?

File: src/postgres/src/backend/tcop/postgres.c [hunk 4]
```diff
@@ -7867,39 +7859,74 @@ disable_statement_timeout(void)
...
  * traceparent_out must point to a buffer of at least
  * YB_TRACEPARENT_VALUE_LEN + 1 bytes.
  *
- * The traceparent must appear in a comment at the start of the query:
+ * The traceparent comments may appear at the query start or end.
+ * If multiple block comments contain a traceparent, the first one is returned.
+ *
  * "/\*traceparent='00-00000000000000000000000000000009-0000000000000005-01'*\/ SELECT 1;"
+ * "SELECT 1 /\*traceparent='00-00000000000000000000000000000009-0000000000000005-01'*\/;"
+
```
Comment (IshanChhangani) [inline src/postgres/src/backend/tcop/postgres.c]:
```suggestion
```

File: src/postgres/src/backend/tcop/postgres.c [hunk 5]
```diff
@@ -7867,39 +7859,74 @@ disable_statement_timeout(void)
...
 /*
- * Extract trace parent from a leading SQL comment in a query.
+ * Extract trace parent from comment in a query.
  *
  * traceparent_out must point to a buffer of at least
  * YB_TRACEPARENT_VALUE_LEN + 1 bytes.
  *
- * The traceparent must appear in a comment at the start of the query:
+ * The traceparent comments may appear at the query start or end.
+ * If multiple block comments contain a traceparent, the first one is returned.
```
Comment (IshanChhangani) [inline src/postgres/src/backend/tcop/postgres.c]: Can we document what happens if a traceparent is given both at the start and end of the query.

File: src/postgres/src/backend/tcop/postgres.c [hunk 6]
```diff
@@ -7867,39 +7859,74 @@ disable_statement_timeout(void)
...
+	const char		   *pos;
+	const char		   *end;
+	const char		   *content;
+	const char		   *comment_end;
+	YbTraceparentResult result;
+
+	/* Check for a leading block comment. */
+	pos = query;
 	while (isspace((unsigned char) *pos))
 		pos++;
```
Comment (IshanChhangani) [inline src/postgres/src/backend/tcop/postgres.c:7878]:
```suggestion
		pos++;
		
```

**Review by IshanChhangani — CHANGES_REQUESTED** [review]

File: src/postgres/src/backend/tcop/postgres.c (hunk 5 above)
Comment (egladysh) [reply src/postgres/src/backend/tcop/postgres.c]: It's kind document it perhaps not very precise: the first from left-to-right. "If multiple block comments contain a traceparent, the first one is returned."

File: src/yb/yql/pgwrapper/dist_trace-test.cc [hunk 7]
```diff
@@ -917,6 +917,52 @@ TEST_F(DistTraceTest, TestAbortSpan) {
...
+    auto tp = GenerateTraceparent();
+    auto query = Format("SELECT 1 /*traceparent='$0'*/", tp.full);
+    ASSERT_OK(conn_->Fetch(query));
+    expected_traces.push_back(MakeExpectedTrace(query, tp.trace_id, kSelectSpans));
+  }
+
+  // Trailing comment with surrounding whitespace.
+  {
+    auto tp = GenerateTraceparent();
+    auto query = Format("SELECT 2 /*traceparent='$0'*/  ", tp.full);
```
Comment (IshanChhangani) [inline src/yb/yql/pgwrapper/dist_trace-test.cc:952]: Liked the trick of changing the select query's constant. This protects us from clubbing spans from multiple traceparents into one. I tried resolving for that by adding a clearTraces function. And calling this function before each query. Though yours is a better solution here.

File: src/yb/yql/pgwrapper/dist_trace-test.cc [hunk 8]
```diff
@@ -917,6 +917,52 @@ TEST_F(DistTraceTest, TestAbortSpan) {
...
+  {
+    auto tp1 = GenerateTraceparent();
+    auto tp2 = GenerateTraceparent();
+    auto query = Format(
+        "/*traceparent='$0'*/ SELECT 3 /*traceparent='$1'*/", tp1.full, tp2.full);
+    ASSERT_OK(conn_->Fetch(query));
+    expected_traces.push_back(MakeExpectedTrace(query, tp1.trace_id, kSelectSpans));
+  }
+
+  // Mid-query comment must NOT produce a span (only leading/trailing are supported).
```
Comment (IshanChhangani) [inline src/yb/yql/pgwrapper/dist_trace-test.cc:967]: This looks good. Can we add some more edge cases that won't be parsed like:
`/* abc */ /*traceparent='<valid>'*/ select 1`
`select 1 /*traceparent='<valid>'*/ /* abc */ `
`/*traceparent='<invalid>'*/ select 1 /*traceparent='<valid>'*/` -> this is a special case and since we prefer leading traceparent over trailing one this should just be giving a warning that traceparent supplied is invalid and execute the query without tracing it.

File: src/postgres/src/backend/tcop/postgres.c [hunk 9]
```diff
@@ -7852,39 +7852,74 @@ disable_statement_timeout(void)
...
-	Assert(comment_end);
+	if (end - query < 2 * YB_TRACEPARENT_COMMENT_DELIMITERS_LEN ||
+		end[-2] != '*' || end[-1] != '/')
+		return YB_TRACEPARENT_NO_COMMENT;
+
+	/* Scan backwards to find the slash-star that opens the trailing comment. */
+	pos = end - YB_TRACEPARENT_COMMENT_DELIMITERS_LEN; /* points to '*' of closing delimiter */
+	while (pos > query)
+	{
+		pos--;
```
Comment (IshanChhangani) [inline src/postgres/src/backend/tcop/postgres.c:7953]:
```suggestion
		pos--;
		
```

File: src/yb/yql/pgwrapper/dist_trace-test.cc (hunk 8 above)
Comment (egladysh) [reply src/yb/yql/pgwrapper/dist_trace-test.cc:967]: done

**Comment by egladysh** [comment]
trigger jenkins

**Comment by IshanChhangani** [comment]
Trigger Jenkins

**Comment by hari90** [comment]
Phorge diff synced: **[D52273](https://phorge.dev.yugabyte.com/D52273)**
Commit: `e1e103df`

Jenkins build has been triggered. Results will be posted here once it completes.

---
*JenkinsBot*

**Comment by hari90** [comment]
✅ **Jenkins build for commit `e1e103df`: Pass**

**Passed:** 12

---
### 🔨 DB Build/Test Job Summary
| Build | Total | Passed | Failed | Failed After Retries |
|-------|------:|-------:|-------:|---------------------:|
| D52273-arm-alma8-clang21-release | 11228 | 10801 | **5** | 1 |
| D52273-alma8-clang21-release | 11230 | 10804 | **5** | 2 |
| D52273-mac14-clang-release | 2 | 2 | 0 | 0 |
| D52273-arm-mac14-clang-release | 17 | 17 | 0 | 0 |
| D52273-ubuntu22.04-clang21-debug | 2 | 2 | 0 | 0 |
| D52273-alma9-clang21-asan | 11847 | 11109 | **16** | 10 |
| D52273-alma8-clang21-tsan | 11755 | 10110 | **15** | 8 |
| D52273-alma8-gcc12-fastdebug | 11970 | 11506 | **7** | 2 |

[Full status](http://csiweb.dev.yugabyte.com/D52273)

---
*JenkinsBot*

**Comment by IshanChhangani** [comment]
LGTM!

**Review by IshanChhangani — APPROVED** [review]
